pubsub/psclient/client.go

```diff
@@ -1,97 +1,347 @@
+		responses:    make(map[int64]*pstypes.ResponseMessage),
+		ctx:          ctx,
+		shutdown:     shutdown,
+	}
+
+	go cl.receiver()
+
+	return cl, nil
 }
 
-// New creates a new Client from a *websocket.Conn.
-func New(ws *websocket.Conn) *Client {
+// NewFromConn creates a new Client from a *websocket.Conn.
+func NewFromConn(ws *websocket.Conn, ctx context.Context, opts *Opts) *Client {
 	if ws == nil {
 		return nil
 	}
-	return &Client{
+
+	readTimeout, writeTimeout := defaultTimeout, defaultTimeout
+	if opts != nil {
+		readTimeout = opts.ReadTimeout
+		writeTimeout = opts.WriteTimeout
+	}
+
+	ctx, shutdown := context.WithCancel(ctx)
+	cl := &Client{
 		Conn:         ws,
-		ReadTimeout:  defaultTimeout,
-		WriteTimeout: defaultTimeout,
+		readTimeout:  readTimeout,
+		writeTimeout: writeTimeout,
+		sendMsgChan:  make(chan string, 16),
+		recvMsgChan:  make(chan *ClientMessage, 16),
+		requests:     make(map[int64]chan *pstypes.ResponseMessage),
+		responses:    make(map[int64]*pstypes.ResponseMessage),
+		ctx:          ctx,
+		shutdown:     shutdown,
+	}
+
+	go cl.receiver()
+
+	return cl
+}
+
+// Stop shutsdown the Client. If the websocket connection was created during
+// Client construction, it is is shutdown too.
+func (c *Client) Stop() {
+	log.Trace("Stopping psclient Client...")
+	// Close the user's receive channel.
+	close(c.recvMsgChan)
+
+	// Cancel the Context to stop the receiver loop.
+	c.shutdown()
+
+	// Close the websocket connection.
+	if c.ourConn {
+		if err := c.Conn.Close(); err != nil {
+			log.Errorf("Failed to Close websocket connection: %v", err)
+		}
 	}
 }
 
+// ClientMessage represents a message for a client connection.  The
+// (*Client).Receive method provides a channel such messages from the server.
+type ClientMessage struct {
+	EventId string
+	Message interface{}
+}
+
+// Receive gets a receive-only *ClientMessage channel, through which all
+// messages from the server to the client should be received.
+func (c *Client) Receive() <-chan *ClientMessage {
+	return c.recvMsgChan
+}
+
+// receiver receives and decodes messages from the server. It is to be run as a
+// goroutine by the constructor of the Client. The decoded ClientMessages are
+// sent on the recvMsgChan, the channel that is obtained via Receive.
+func (c *Client) receiver() {
+	defer c.shutdown()
+
+	for {
+		if c.ctx.Err() != nil {
+			log.Trace("receiver: context canceled...")
+			return
+		}
+
+		resp, err := c.receiveMsg()
+		if err != nil {
+			if pstypes.IsIOTimeoutErr(err) {
+				continue
+			}
+			log.Errorf("ReceiveMsg failed: %v", err)
+			return
+		}
+
+		msg, err := DecodeMsg(resp)
+		if err != nil {
+			log.Errorf("Failed to decode message: %v", err)
+			continue
+		}
+
+		switch m := msg.(type) {
+		case *pstypes.ResponseMessage:
+			log.Debugf("Response to %s request ID=%d received. Success = %v. Data: %v",
+				m.RequestEventId, m.RequestId, m.Success, m.Data)
+			respChan := c.responseChan(m.RequestId)
```

Comment:
A response message has a unique (for this client) request ID.  In this case, the response is sent to the associated `respChan`.